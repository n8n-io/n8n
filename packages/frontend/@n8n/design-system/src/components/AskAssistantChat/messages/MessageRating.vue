<script setup lang="ts">
import { ref } from 'vue';

import { useI18n } from '@n8n/design-system/composables/useI18n';
import type { RatingFeedback } from '@n8n/design-system/types';

import N8nButton from '../../N8nButton';
import N8nIconButton from '../../N8nIconButton';
import N8nInput from '../../N8nInput';

interface Props {
	style?: 'regular' | 'minimal';
	showFeedback?: boolean;
}

const props = withDefaults(defineProps<Props>(), {
	style: 'regular',
	showFeedback: true,
});

const emit = defineEmits<{
	feedback: [RatingFeedback];
}>();

const { t } = useI18n();
const showRatingButtons = ref(true);
const showFeedbackArea = ref(false);
const showSuccess = ref(false);
const selectedRating = ref<'up' | 'down' | null>(null);
const feedback = ref('');

function onRateButton(rating: 'up' | 'down') {
	selectedRating.value = rating;
	showRatingButtons.value = false;

	emit('feedback', { rating });
	if (props.showFeedback) {
		showFeedbackArea.value = true;
	} else {
		showSuccess.value = true;
	}
}

function onSubmitFeedback() {
	if (selectedRating.value) {
		emit('feedback', { feedback: feedback.value });
		showFeedbackArea.value = false;
		showSuccess.value = true;
	}
}

function onCancelFeedback() {
	showFeedbackArea.value = false;
	showRatingButtons.value = true;
	selectedRating.value = null;
	feedback.value = '';
}
</script>

<template>
	<div :class="[$style.rating, $style[style]]">
		<div v-if="showRatingButtons" :class="$style.buttons">
			<template v-if="style === 'regular'">
				<N8nButton
					type="secondary"
					size="small"
					:label="t('assistantChat.builder.thumbsUp')"
					data-test-id="message-thumbs-up-button"
					icon="thumbs-up"
					@click="onRateButton('up')"
				/>
				<N8nButton
					type="secondary"
					size="small"
					data-test-id="message-thumbs-down-button"
					:label="t('assistantChat.builder.thumbsDown')"
					icon="thumbs-down"
					@click="onRateButton('down')"
				/>
			</template>
			<template v-else>
				<N8nIconButton
					type="tertiary"
					size="small"
					text
					icon="thumbs-up"
					data-test-id="message-thumbs-up-button"
					@click="onRateButton('up')"
				/>
				<N8nIconButton
					type="tertiary"
					size="small"
					text
					icon="thumbs-down"
					data-test-id="message-thumbs-down-button"
					@click="onRateButton('down')"
				/>
			</template>
		</div>

		<div v-if="showFeedbackArea" :class="$style.feedbackContainer">
			<N8nInput
				v-model="feedback"
				:class="$style.feedbackInput"
				type="textarea"
				:placeholder="t('assistantChat.builder.feedbackPlaceholder')"
				data-test-id="message-feedback-input"
				:read-only="false"
				resize="none"
				:rows="style === 'minimal' ? 3 : 5"
			/>
			<div :class="$style.feedbackActions">
				<N8nButton
					type="secondary"
					size="small"
					:label="t('generic.cancel')"
					@click="onCancelFeedback"
				/>
				<N8nButton
					type="primary"
					size="small"
					data-test-id="message-submit-feedback-button"
					:label="t('assistantChat.builder.submit')"
					@click="onSubmitFeedback"
				/>
			</div>
		</div>

		<p v-if="showSuccess" :class="$style.success">
			{{ t('assistantChat.builder.success') }}
		</p>
	</div>
</template>

<style lang="scss" module>
.rating {
	display: flex;
	flex-direction: column;
	gap: var(--spacing-2xs);
	margin-top: var(--spacing-2xs);
}

.buttons {
	display: flex;
	gap: var(--spacing-2xs);
}

.feedbackContainer {
	display: flex;
	flex-direction: column;
	gap: var(--spacing-2xs);
}

.feedbackInput {
	:global(.el-textarea__inner) {
		resize: none;
		font-family: var(--font-family);
		font-size: var(--font-size-2xs);
	}
}

.feedbackActions {
	display: flex;
	gap: var(--spacing-2xs);
	justify-content: flex-end;
}

.success {
	color: var(--color-success);
	font-size: var(--font-size-2xs);
	margin: 0;
}

/* Minimal style specific */
.minimal {
	margin-top: 0;
	.buttons {
		gap: var(--spacing-3xs);
	}

	.feedbackContainer {
		gap: var(--spacing-3xs);
	}

	.feedbackInput {
		:global(.el-textarea__inner) {
			font-size: var(--font-size-3xs);
		}
	}
}
</style>
