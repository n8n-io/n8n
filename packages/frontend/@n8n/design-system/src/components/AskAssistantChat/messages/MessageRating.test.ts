import { render, fireEvent, waitFor } from '@testing-library/vue';
import { createPinia, setActivePinia } from 'pinia';
import { nextTick } from 'vue';

import MessageRating from './MessageRating.vue';

const stubs = ['n8n-button', 'n8n-icon-button', 'n8n-input'];

// Mock i18n to return keys instead of translated text
vi.mock('@n8n/design-system/composables/useI18n', () => ({
	useI18n: () => ({
		t: (key: string) => key,
	}),
}));

beforeEach(() => {
	setActivePinia(createPinia());
});

describe('MessageRating', () => {
	it('should render correctly with default props', () => {
		const wrapper = render(MessageRating, {
			global: { stubs },
		});

		expect(
			wrapper.container.querySelector('[data-test-id="message-thumbs-up-button"]'),
		).toBeTruthy();
		expect(
			wrapper.container.querySelector('[data-test-id="message-thumbs-down-button"]'),
		).toBeTruthy();
		expect(wrapper.html()).toMatchSnapshot();
	});

	describe('rating interactions', () => {
		it('should emit feedback when thumbs up is clicked', async () => {
			const wrapper = render(MessageRating, {
				global: { stubs },
			});

			const upButton = wrapper.container.querySelector('[data-test-id="message-thumbs-up-button"]');
			await fireEvent.click(upButton!);

			expect(wrapper.emitted()).toHaveProperty('feedback');
			expect(wrapper.emitted().feedback[0]).toEqual([{ rating: 'up' }]);
		});

		it('should emit feedback when thumbs down is clicked', async () => {
			const wrapper = render(MessageRating, {
				global: { stubs },
			});

			const downButton = wrapper.container.querySelector(
				'[data-test-id="message-thumbs-down-button"]',
			);
			await fireEvent.click(downButton!);

			expect(wrapper.emitted()).toHaveProperty('feedback');
			expect(wrapper.emitted().feedback[0]).toEqual([{ rating: 'down' }]);
		});

		it('should hide rating buttons and show success after thumbs up', async () => {
			const wrapper = render(MessageRating, {
				global: { stubs },
			});

			const upButton = wrapper.container.querySelector('[data-test-id="message-thumbs-up-button"]');
			await fireEvent.click(upButton!);
			await nextTick();

			expect(
				wrapper.container.querySelector('[data-test-id="message-thumbs-up-button"]'),
			).toBeFalsy();
			expect(
				wrapper.container.querySelector('[data-test-id="message-thumbs-down-button"]'),
			).toBeFalsy();
			expect(wrapper.getByText('assistantChat.builder.success')).toBeTruthy();
		});

		it('should hide rating buttons and show feedback form after thumbs down when showFeedback is true', async () => {
			const wrapper = render(MessageRating, {
				props: { showFeedback: true },
				global: { stubs },
			});

			const downButton = wrapper.container.querySelector(
				'[data-test-id="message-thumbs-down-button"]',
			);
			await fireEvent.click(downButton!);
			await nextTick();

			expect(
				wrapper.container.querySelector('[data-test-id="message-thumbs-up-button"]'),
			).toBeFalsy();
			expect(
				wrapper.container.querySelector('[data-test-id="message-thumbs-down-button"]'),
			).toBeFalsy();
			expect(
				wrapper.container.querySelector('[data-test-id="message-feedback-input"]'),
			).toBeTruthy();
			expect(
				wrapper.container.querySelector('[data-test-id="message-submit-feedback-button"]'),
			).toBeTruthy();
		});

		it('should show success immediately after thumbs down when showFeedback is false', async () => {
			const wrapper = render(MessageRating, {
				props: { showFeedback: false },
				global: { stubs },
			});

			const downButton = wrapper.container.querySelector(
				'[data-test-id="message-thumbs-down-button"]',
			);
			await fireEvent.click(downButton!);
			await nextTick();

			expect(
				wrapper.container.querySelector('[data-test-id="message-feedback-input"]'),
			).toBeFalsy();
			expect(wrapper.getByText('assistantChat.builder.success')).toBeTruthy();
		});
	});

	describe('feedback form interactions', () => {
		it('should submit feedback and show success', async () => {
			const wrapper = render(MessageRating, {
				props: { showFeedback: true },
				global: { stubs: ['n8n-button', 'n8n-icon-button'] }, // Don't stub n8n-input
			});

			const downButton = wrapper.container.querySelector(
				'[data-test-id="message-thumbs-down-button"]',
			);
			await fireEvent.click(downButton!);
			await nextTick();

			// Find the actual textarea element within the N8nInput component
			const textarea = wrapper.container.querySelector(
				'textarea[data-test-id="message-feedback-input"]',
			);
			await fireEvent.update(textarea!, 'This is my feedback about the response');
			await nextTick();

			const submitButton = wrapper.container.querySelector(
				'[data-test-id="message-submit-feedback-button"]',
			);
			await fireEvent.click(submitButton!);
			await nextTick();

			expect(wrapper.emitted().feedback).toHaveLength(2);
			expect(wrapper.emitted().feedback[1]).toEqual([
				{ feedback: 'This is my feedback about the response' },
			]);
			expect(
				wrapper.container.querySelector('[data-test-id="message-feedback-input"]'),
			).toBeFalsy();
			expect(wrapper.getByText('assistantChat.builder.success')).toBeTruthy();
		});

		it('should cancel feedback and return to rating buttons', async () => {
			const wrapper = render(MessageRating, {
				props: { showFeedback: true },
				global: { stubs },
			});

			const downButton = wrapper.container.querySelector(
				'[data-test-id="message-thumbs-down-button"]',
			);
			await fireEvent.click(downButton!);
			await nextTick();

			const cancelButton = wrapper.container.querySelector(
				'n8n-button-stub[label="generic.cancel"]',
			);
			await fireEvent.click(cancelButton!);
			await nextTick();

			expect(
				wrapper.container.querySelector('[data-test-id="message-feedback-input"]'),
			).toBeFalsy();
			expect(
				wrapper.container.querySelector('[data-test-id="message-thumbs-up-button"]'),
			).toBeTruthy();
			expect(
				wrapper.container.querySelector('[data-test-id="message-thumbs-down-button"]'),
			).toBeTruthy();
		});

		it('should focus feedback input after thumbs down in regular mode', async () => {
			const wrapper = render(MessageRating, {
				props: { showFeedback: true, style: 'regular' },
				global: { stubs },
			});

			const downButton = wrapper.container.querySelector(
				'[data-test-id="message-thumbs-down-button"]',
			);
			await fireEvent.click(downButton!);

			await waitFor(() => {
				const feedbackInput = wrapper.container.querySelector(
					'[data-test-id="message-feedback-input"]',
				);
				expect(feedbackInput).toBeTruthy();
			});
		});

		it('should clear feedback text when cancelling', async () => {
			const wrapper = render(MessageRating, {
				props: { showFeedback: true },
				global: { stubs },
			});

			const downButton = wrapper.container.querySelector(
				'[data-test-id="message-thumbs-down-button"]',
			);
			await fireEvent.click(downButton!);
			await nextTick();

			const cancelButton = wrapper.container.querySelector(
				'n8n-button-stub[label="generic.cancel"]',
			);
			await fireEvent.click(cancelButton!);
			await nextTick();

			const downButtonAgain = wrapper.container.querySelector(
				'[data-test-id="message-thumbs-down-button"]',
			);
			await fireEvent.click(downButtonAgain!);
			await nextTick();

			const feedbackInputAfter = wrapper.container.querySelector(
				'[data-test-id="message-feedback-input"]',
			);
			expect(feedbackInputAfter?.getAttribute('modelvalue')).toBe('');
		});
	});

	describe('textarea rows based on style', () => {
		it('should have 5 rows for regular style', async () => {
			const wrapper = render(MessageRating, {
				props: { showFeedback: true, style: 'regular' },
				global: { stubs },
			});

			const downButton = wrapper.container.querySelector(
				'[data-test-id="message-thumbs-down-button"]',
			);
			await fireEvent.click(downButton!);
			await nextTick();

			const feedbackInput = wrapper.container.querySelector(
				'[data-test-id="message-feedback-input"]',
			);
			expect(feedbackInput?.getAttribute('rows')).toBe('5');
		});

		it('should have 3 rows for minimal style', async () => {
			const wrapper = render(MessageRating, {
				props: { showFeedback: true, style: 'minimal' },
				global: { stubs },
			});

			const downButton = wrapper.container.querySelector(
				'[data-test-id="message-thumbs-down-button"]',
			);
			await fireEvent.click(downButton!);
			await nextTick();

			const feedbackInput = wrapper.container.querySelector(
				'[data-test-id="message-feedback-input"]',
			);
			expect(feedbackInput?.getAttribute('rows')).toBe('3');
		});
	});
});
