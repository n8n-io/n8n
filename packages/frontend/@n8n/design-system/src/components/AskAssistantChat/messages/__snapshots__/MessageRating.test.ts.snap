// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`MessageRating > should render correctly with default props 1`] = `
"<div class="rating regular">
  <div class="buttons">
    <n8n-button-stub icon="thumbs-up" block="false" element="button" label="assistantChat.builder.thumbsUp" square="false" active="false" disabled="false" loading="false" outline="false" size="small" text="false" type="secondary" data-test-id="message-thumbs-up-button"></n8n-button-stub>
    <n8n-button-stub icon="thumbs-down" block="false" element="button" label="assistantChat.builder.thumbsDown" square="false" active="false" disabled="false" loading="false" outline="false" size="small" text="false" type="secondary" data-test-id="message-thumbs-down-button"></n8n-button-stub>
  </div>
  <!--v-if-->
  <!--v-if-->
</div>"
`;
